### ISSUE TITLE
Plugin incompatible with AS 4.0.1

### ISSUE THREAD

**Issue by rakesh-sh2020** [facebook/ktfmt#61]

**Comment by jef** [comment]
https://github.com/facebookincubator/ktfmt/pull/62

**Comment by jef** [comment]
I believe AS `4.0.1` is build `193.6626763`, which should've been covered in the current plugin version `2020.2` (`202.7660.17`).

That being said #62, may resolve this issue. Could use some testing around it. They don't have a value for AS in https://github.com/JetBrains/gradle-intellij-plugin#intellij-platform-properties, so could be somewhat manual test.

**Comment by cgrushko** [comment]
I'll publish a new plugin shortly.

**Comment by cgrushko** [comment]
Done (currently under JetBrains review)

**Comment by somesh-basicx** [comment]
@cgrushko Any update on this... I still can't see any plugin on their marketplace.

**Comment by cgrushko** [comment]
@somesh-basicx the latest version (1.1.0.18) shows it's compatible with "Android Studio — build 203.4203 — 203.*"
(https://plugins.jetbrains.com/plugin/14912-ktfmt/versions)

I guess that's too new for 193?

Sorry it's such a mess, I haven't yet figured out how to release the plugin properly, and haven't had the cycles to integrate-test the plugin with multiple IntelliJ versions.

Would you be willing to try to build a local copy after playing with the build settings and finding what's needed for AS 4.0.2?

**Comment by jef** [comment]
> Sorry it's such a mess, I haven't yet figured out how to release the plugin properly, and haven't had the cycles to integrate-test the plugin with multiple IntelliJ versions.

https://github.com/facebookincubator/ktfmt/pull/62#issuecomment-705065362

> I guess that's too new for 193?

> All is well. I think I understand the `sinceBuild` now. Probably doesn't become available until 203*. Not sure if it's needed really. I would be interested in the side effects of removing.

I'm thinking we remove `sinceBuild` and it would resolve this issue. I can download AS tonight and build it.

**Comment by TheSomeshKumar** [comment]
@cgrushko Okay, I'll see what I can do here. 

Just want to mention that Android Studio 4.0.2 (Stable) is version 193.* & we just got a new stable update which is AS 4.1 (version 201.*)  and the AS 4.2 in the canary branch is 202.* 

So I think Android developers won't get version 203 until 2021 Q2. So we need to make this plugin compatible with current AS release.

**Comment by jef** [comment]
Definitely the case. Thanks for the report @TheSomeshKumar. I'll carve in some time tonight.

**Comment by alzahm** [comment]
Any update on this? 
I am having a problem installing it on AS 4.1.1.

**Comment by JavierSegoviaCordoba** [comment]
After checking [this post](https://intellij-support.jetbrains.com/hc/en-us/community/posts/360003338799/comments/360001685699) I think we need to use:

```
intellij {
    version = "2019.3.4"
}

patchPluginXml {
    sinceBuild("193")
    untilBuild("")
}
```

**Comment by cgrushko** [comment]
I tried making the plugin work with AS 4.0.1., which is v193, and unfortunately there's too many things to change to make it work.
I uploaded a version of the plugin to JetBrains that runs on v201 and up (Android 4.1, IntelliJ 2020.1 and later) and uses ktfmt 0.19. 
If you can't upgrade to Android 4.1, I suggest using the Spotless intergration.

Supporting AS 4.0 is probably a non-trivial amount of work at this point. If anyone's interesting in picking it up, I suggest looking at google-java-format's plugin code at their v1.7 tag.

I'll close this Issue for now, unless anyone thinks it's worth keeping open for documentation purposes?

**Comment by TheSomeshKumar** [comment]
@cgrushko  Using Android Studio 4.1.1 (Build #AI-201.8743.12.41.6953283) but still can't find ktfmt plugin in the plugin marketplace of Studio 😟

**Comment by cgrushko** [comment]
@TheSomeshKumar the update is stuck in review - they say there's a problem with the compatibility settings and I'm following up with them.

Meanwhile, try downloading from https://plugins.jetbrains.com/plugin/download?rel=true&updateId=106669 (that's the zip file I uploaded to them) and installing from disk. 
Alternatively, clone the repo and run `./gradlew buildPlugin` inside the ktfmt_idea_plugin subdirectory. Then, find the resulting .zip file in the build/ subdirectory and install it from disk.

**Comment by TheSomeshKumar** [comment]
@cgrushko Thanks for sharing the plugin. After loading the plugin from disk Android Studio doesn't show me any settings to enable the ktfmt plugin anywhere but when I installed the same zip file in the IntelliJ 2020.3 it is showing me appropriate settings.

**Comment by cgrushko** [comment]
Ugh, 
```
com.intellij.diagnostic.PluginException: While loading class com.facebook.ktfmt.intellij.KtfmtConfigurable: com/facebook/ktfmt/intellij/KtfmtConfigurable has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0 [Plugin: com.facebook.ktfmt_idea_plugin] [Plugin: com.facebook.ktfmt_idea_plugin]
```

This means that ktfmt_idea_plugin was compiled with Java 11 but AS 4.1.1's JRE is Java 8.
We're using 11 because we depend on google-java-format 1.8 which requires Java 11.

Interestingly, the google-java-format has the same problem: it fails to load with the same error.

I'm not sure what to do at this point. I think it might be worth for someone to rewrite the plugin from scratch, possibly in another repo.

**Comment by JavierSegoviaCordoba** [comment]
I think 4.2 stable will be released in 2021 Q1 (February = 6 months since 4.1.0), maybe just wait for it.

**Comment by cgrushko** [comment]
I found google-java-format's parallel issue: https://github.com/google/google-java-format/issues/536

**Comment by JavierSegoviaCordoba** [comment]
Have you tried this? If it works the current plugin state should work with that
 https://github.com/google/google-java-format/issues/536#issuecomment-716517299

**Comment by cgrushko** [comment]
Ok, so that works:
AS doesn't show Choose Runtime in its Plugin Marketplace. I was able to download it, install it from disk, and point it at a JetBrains JDK 11.

I'll put it in the release notes, and ask JetBrains to release the update.

tl;dr - the ktfmt plugin works with v201+, which includes Android Studio 4.1.1, as long as it's running under JDK 11.
This is a limitation coming from google-java-format. We can fix GJF to use JDK 8 again.

**Comment by TheSomeshKumar** [comment]
Android Studio 4.2 Beta 2 is now available and there will be a **R**elease **C**andidate sooner than we think... Usually, most android developers update Android Studio as soon as it is available so I don't think supporting version 4.1 is worth the time & efforts.

**Comment by cgrushko** [comment]
The limiting factor at this point is the Java version under which IntelliJ
/ AS is running.
IIUC, AS 4.2 is based on v202, which doesn’t necessarily run on Java 11.

**Comment by JavierSegoviaCordoba** [comment]
@cgrushko but I am using ktfmt with AS 4.2 beta with no problems, so it should be using Java 11 I think

**Comment by cgrushko** [comment]
You can verify by going to About. It should say which runtime is used under
the platform version.

**Comment by JavierSegoviaCordoba** [comment]
On macOS -> Runtime version: 11.0.8+10

**Comment by TheSomeshKumar** [comment]
Android Studio 4.2 will use Java 11 [https://developer.android.com/studio/preview/features/#optimizing_gradle_builds_for_jdk_11]()

**Comment by hick209** [comment]
Is this still an issue?

Since it's over 2 years old I'm closing it, but please do reopen in case it is still an issue
